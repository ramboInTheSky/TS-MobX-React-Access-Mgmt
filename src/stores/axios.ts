import axios from 'axios'

// tslint:disable
axios.interceptors &&
    (function() {
        // Add a request interceptor
        axios.interceptors.request.use(
            function(config) {
                // Do something before request is sent
                return config
            },
            function(error) {
                // Do something with request error
                return Promise.reject(error)
            }
        )

        // Add a response interceptor
        axios.interceptors.response.use(
            function(response) {
                // Do something with response data
                return response
            },
            function(error) {
                console.debug('Handling error response')
                if (error.response && error.response.status === 403) {
                    console.debug('handling 403')
                }
                return Promise.reject(error)
            }
        )
    })()

export default axios
