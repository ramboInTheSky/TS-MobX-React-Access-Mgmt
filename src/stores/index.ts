// first party
import { ProfileStore } from './Profiles'
import { Routing } from './Routing'
import { VisitStore } from './Visits'
import { LockGroupStore } from './LockGroups'
import { AuthStore } from './Auth.Code.Grant'
import { NotificationStore } from './Notifications'

// re-exports
export * from './Profiles'
export * from './Routing'
export * from './Visits'
export * from './LockGroups'
export * from './Auth.Code.Grant'
export * from './Notifications'

export interface Stores {
    profiles: ProfileStore
    routing: Routing
    visits: VisitStore
    lockGroups: LockGroupStore
    auth: AuthStore
    notifications: NotificationStore
}

type StoreNamesGuard = keyof Stores

export class StoreNames {
    public static profiles: StoreNamesGuard = 'profiles'
    public static notifications: StoreNamesGuard = 'notifications'
    public static routing: StoreNamesGuard = 'routing'
    public static visits: StoreNamesGuard = 'visits'
    public static lockGroups: StoreNamesGuard = 'lockGroups'
    public static auth: StoreNamesGuard = 'auth'
}

const localStores: Partial<Stores> = {
    profiles: new ProfileStore(),
    routing: new Routing(),
    visits: new VisitStore(),
    lockGroups: new LockGroupStore(),
    auth: new AuthStore(),
    notifications: new NotificationStore(),
}

// test comment
export const stores = { ...localStores }
