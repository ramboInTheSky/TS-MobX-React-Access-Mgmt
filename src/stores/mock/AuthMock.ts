export class AuthMock {
    public user: any = {
        accessToken: 'xZXnpwklGhqd-zN5pJKbkHXNg5jBdJYu',
        idToken:
            'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik5UY3â€¦XhKpVLcBSw6qUe4KgjCI4sIgDa3_Kh7TbaSLgd__pJdZLriQg',
        idTokenPayload: {},
        appState: null,
        refreshToken: null,
    }
    public login = jest.fn()
    public checkSession = jest.fn()
    public logout = jest.fn()
    public getUser = jest.fn(() => this.user)
    public redirectToPage = '/profile/123'
}
