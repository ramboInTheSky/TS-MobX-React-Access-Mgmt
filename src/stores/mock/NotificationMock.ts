export class NotificationMock {
    public setNotification: Function = jest.fn()
}
