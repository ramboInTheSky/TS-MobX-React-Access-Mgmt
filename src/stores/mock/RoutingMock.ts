import { Routing } from '../Routing'

export class RoutingMock extends Routing {
    public goToPage: any = jest.fn()
    public goBackToPage: any = jest.fn()
    public push: any = jest.fn()
    public replace: any = jest.fn()
    public go: any = jest.fn()
    public goBack: any = jest.fn()
    public goForward: any = jest.fn()
    public breadcrumb: any[] = [{ page: 'profile', id: '123-123-123' }]
    public location: any = {}
    public history: any = {}
    public getBreadcrumb: any = jest.fn(() => this.breadcrumb)
    public resizeBreadcrumb: any = jest.fn()
    public resizeBreadcrumbs: any = jest.fn()
    public splitThings: any = jest.fn()
    public rebuildBreadcrumb: any = jest.fn()
}
