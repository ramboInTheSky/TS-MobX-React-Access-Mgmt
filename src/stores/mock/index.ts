export * from './RoutingMock'
export * from './ProfilesMock'
export * from './VisitsMock'
export * from './AuthMock'
export * from './NotificationMock'

export const mockDateConstructor = () => {
    const now = new Date('2020')
    /*eslint no-global-assign:off*/
    Date = class extends Date {
        constructor() {
            super()
            return now
        }
    } as DateConstructor
}
