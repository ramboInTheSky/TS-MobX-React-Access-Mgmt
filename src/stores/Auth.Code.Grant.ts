import * as Sentry from '@sentry/browser'
import { Auth0UserProfile, WebAuth, AuthOptions } from 'auth0-js'
import { observable } from 'mobx'
import axios from './axios'

import opts from '../constants/auth'
import {
    getConfig,
    randomString,
    sha256,
    bufferToBase64UrlEncoded,
} from '../utils/authUtils'

const REFRESH_TOKEN_INTERVAL = 30 // seconds
const LOGOUT_INTERVAL = 5 * 60 // seconds

export class AuthStore {
    private logoutTimeout: any = null
    private checkSessionTimeout: any = null

    private auth0: WebAuth

    @observable
    private loggedUser?: Auth0UserProfile

    @observable
    private redirect?: string

    private settings: AuthOptions

    private config: any

    constructor() {
        this.settings = {
            domain: opts.AUTH0_DOMAIN,
            clientID: opts.AUTH0_CLIENT_ID,
            responseType: opts.AUTH0_RESPONSE_TYPE,
            redirectUri: opts.REDIRECT_URI,
        }
        this.auth0 = new WebAuth(this.settings)
    }

    // public api //

    public get user() {
        return this.loggedUser
    }

    public get redirectToPage() {
        return this.redirect
    }

    // TODO check security of this!
    get auth0Instance() {
        return this.auth0
    }

    public resetTimer() {
        this.setTimer()
    }

    public async login() {
        try {
            this.config = this.config || (await getConfig(opts.AUTH0_DOMAIN))
            const search = new URLSearchParams(window.location.search)

            if (!search.has('code')) {
                this.redirectToLoginPage()
            }

            const code = search.get('code')
            const state = search.get('state')
            const code_verifier = sessionStorage.getItem(
                `login-code-verifier-${state}`
            )
            sessionStorage.removeItem(`login-code-verifier-${state}`)
            this.redirect = localStorage.getItem('currentPage') || undefined
            window.history.pushState('', document.title, window.location.origin)

            if (!code_verifier) {
                console.debug('unexpected state parameter')
                return
            }

            const settings: any = {
                code_verifier,
                code,
                audience: opts.API_AUDIENCE,
                client_id: opts.AUTH0_CLIENT_ID,
                redirect_uri: opts.REDIRECT_URI,
                grant_type: 'authorization_code',
            }

            // exchange the authorization code for a tokenset
            this.exchangeForToken(settings, true)
        } catch (err) {
            console.debug(err)
        }
    }

    public logout = () => {
        this.loggedUser = undefined
        if (this.logoutTimeout) clearTimeout(this.logoutTimeout)
        if (this.checkSessionTimeout) clearTimeout(this.checkSessionTimeout)
        this.auth0.logout({ returnTo: opts.REDIRECT_URI })
    }

    // private api //

    private checkSession = async () => {
        try {
            this.config = this.config || (await getConfig(opts.AUTH0_DOMAIN))
            const state = randomString(32)
            const codeVerifier = randomString(32)
            const codeChallenge = await sha256(codeVerifier).then(
                bufferToBase64UrlEncoded
            )
            const authorizationEndpointUrl = new URL(
                this.config.authorization_endpoint
            )

            // here we encode the authorization request
            authorizationEndpointUrl.search = new URLSearchParams({
                state,
                audience: opts.API_AUDIENCE,
                redirect_uri: opts.REDIRECT_URI,
                client_id: opts.AUTH0_CLIENT_ID,
                response_type: opts.AUTH0_RESPONSE_TYPE,
                response_mode: 'web_message',
                prompt: 'none',
                scope: opts.AUTH0_SCOPE,
                code_challenge: codeChallenge,
                code_challenge_method: 'S256',
            }) as any

            // load the url in an iframe and wait for the response
            let authorizeResponse: any = {}
            try {
                authorizeResponse = await new Promise((resolve, reject) =>
                    this.injectIphrame(
                        resolve,
                        reject,
                        authorizationEndpointUrl,
                        state
                    )
                )
            } catch (injectIphrameError) {
                this.logout()
                console.debug(injectIphrameError)
            }

            // exchange the authorization code for a tokenset
            const settings = {
                audience: opts.API_AUDIENCE,
                client_id: opts.AUTH0_CLIENT_ID,
                redirect_uri: opts.REDIRECT_URI,
                grant_type: 'authorization_code',
                code_verifier: codeVerifier,
                code: authorizeResponse.code,
            }
            this.exchangeForToken(settings)
        } catch (err) {
            this.logout()
            console.debug(err)
        }
    }

    private redirectToLoginPage = async () => {
        try {
            localStorage.setItem('currentPage', window.location.pathname)
            const state = randomString(32)
            const codeVerifier = randomString(32)
            const codeChallenge = await sha256(codeVerifier).then(
                bufferToBase64UrlEncoded
            )
            // we need to store the state to validate the callback
            // and also the code verifier to send later
            sessionStorage.setItem(`login-code-verifier-${state}`, codeVerifier)
            this.config = this.config || (await getConfig(opts.AUTH0_DOMAIN))
            const authorizationEndpointUrl = new URL(
                this.config.authorization_endpoint
            )
            // here we encode the authorization request
            authorizationEndpointUrl.search = new URLSearchParams({
                state,
                audience: opts.API_AUDIENCE,
                redirect_uri: opts.REDIRECT_URI,
                // tslint:disable-next-line: no-string-literal
                client_id: opts.AUTH0_CLIENT_ID,
                response_type: 'code',
                scope: opts.AUTH0_SCOPE,
                code_challenge: codeChallenge,
                code_challenge_method: 'S256',
            }) as any

            window.location.assign(authorizationEndpointUrl as any)
        } catch (err) {
            console.debug(err)
        }
    }

    private async exchangeForToken(
        settings: Record<string, string>,
        callUserInfo?: boolean
    ) {
        const tokenSet = await axios.post(
            this.config.token_endpoint,
            new URLSearchParams(settings)
        )

        // this has access_token, id_token, expires_in, scope, token_type.
        if (tokenSet.data.access_token) {
            const exp = tokenSet.data.expires_in * 1000 + new Date().getTime()
            console.debug('session expires at:', new Date(exp))
            this.checkSessionTimeout = setTimeout(
                this.checkSession,
                REFRESH_TOKEN_INTERVAL * 1000
            )
            this.setToken(tokenSet.data)
            if (callUserInfo) {
                try {
                    await this.auth0!.client.userInfo(
                        tokenSet.data.access_token,
                        (error, data) => {
                            if (error) {
                                console.debug(error)
                            }
                            this.setUser(data)
                            this.setTimer()
                        }
                    )
                } catch (userInfoError) {
                    console.debug(userInfoError)
                }
            }
        }
    }

    private setUser(user: Auth0UserProfile) {
        this.loggedUser = user
        Sentry.configureScope(scope => {
            scope.setUser({ username: user.sub })
        })
    }

    private setToken(data: any) {
        if (data.id_token && data.expires_in) {
            try {
                axios.defaults.headers.common.Authorization = `Bearer ${
                    data.access_token
                }` // for all requests
                const axiosInterceptorResponse: any =
                    axios.interceptors.response
                if (axiosInterceptorResponse.handlers.length <= 1) {
                    axios.interceptors.response.use(
                        response => response,
                        error => {
                            if (
                                error.response &&
                                error.response.status === 401
                            ) {
                                console.debug('handling 401 response')
                                this.logout()
                                return error.response
                            }
                            return Promise.reject(error)
                        }
                    )
                }
            } catch (err) {
                console.log('impossible to parse Bearer Token')
            }
        }
    }

    private setTimer() {
        console.debug('logout timeout: ', this.logoutTimeout)
        clearTimeout(this.logoutTimeout)
        this.logoutTimeout = setTimeout(() => {
            this.logout()
        }, LOGOUT_INTERVAL * 1000)
    }

    private injectIphrame = async (
        resolve: any,
        reject: any,
        authorizationEndpointUrl: any,
        state: any
    ) => {
        const iframe = document.createElement('iframe')
        iframe.style.display = 'none'

        const timeoutSetTimeoutId = setTimeout(() => {
            try {
                reject(new Error('timed out'))
            } catch (err) {
                console.debug(err)
            }
            window.document.body.removeChild(iframe)
        }, 60 * 1000)

        function responseHandler(e: any) {
            try {
                if (
                    e.origin !== authorizationEndpointUrl.origin ||
                    e.data.type !== 'authorization_response'
                ) {
                    return
                }
                e.source.close()
                clearTimeout(timeoutSetTimeoutId)
                window.removeEventListener('message', responseHandler, false)
                window.document.body.removeChild(iframe)
                const response = e.data.response
                if (response.error) {
                    return reject(response)
                }
                if (response.state !== state) {
                    return reject(new Error('State does not match.'))
                }
                resolve(response)
            } catch (genericError) {
                console.debug(genericError)
            }
        }

        window.addEventListener('message', responseHandler)
        window.document.body.appendChild(iframe)
        iframe.setAttribute('src', authorizationEndpointUrl as any)
    }
}
export default AuthStore
