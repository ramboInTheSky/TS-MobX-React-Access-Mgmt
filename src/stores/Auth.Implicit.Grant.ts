import { Auth0UserProfile, WebAuth, AuthOptions } from 'auth0-js'
import { observable } from 'mobx'
import opts from '../constants/auth'
import axios from './axios'

const REFRESH_TOKEN_INTERVAL = 60 // seconds
const LOGOUT_INTERVAL = 5 * 60 // seconds

export class AuthStore {
    public get user() {
        return this.loggedUser
    }

    get auth0Instance() {
        return this.auth0
    }

    public get redirectToPage() {
        return this.redirect
    }

    @observable
    private redirect?: string

    @observable
    private auth0: WebAuth
    // private accessToken?: string
    private idToken?: string
    private expiresAt?: number
    private loggedUser?: Auth0UserProfile
    private checkSessionInterval?: NodeJS.Timeout
    private settings: AuthOptions
    private logoutTimeout: any = null

    constructor() {
        this.settings = {
            domain: opts.AUTH0_DOMAIN,
            clientID: opts.AUTH0_CLIENT_ID,
            responseType: opts.AUTH0_RESPONSE_TYPE,
            redirectUri: opts.REDIRECT_URI,
        }
        this.auth0 = new WebAuth(this.settings)
        const userString = localStorage.getItem('user')
        const idTokenString = localStorage.getItem('idToken')
        const expiresAtString = localStorage.getItem('expiresAt')
        this.loggedUser = userString ? JSON.parse(userString) : undefined
        this.idToken = idTokenString ? JSON.parse(idTokenString) : undefined
        this.expiresAt = expiresAtString
            ? JSON.parse(expiresAtString)
            : undefined

        if (this.loggedUser) {
            this.checkSessionInterval = setInterval(
                this.checkSession,
                REFRESH_TOKEN_INTERVAL * 1000
            )
        }
    }

    public getUser = async () => {
        try {
            await this.auth0.parseHash(
                { hash: window.location.hash },
                async (err, data) => {
                    if (err) {
                        return console.log(err)
                    }
                    if (data) {
                        if (data.accessToken) {
                            this.setVars(data)
                            try {
                                return await this.auth0!.client.userInfo(
                                    data.accessToken,
                                    (error, user) => {
                                        if (error) {
                                            console.log(error)
                                        }
                                        this.setUser(user)
                                        return user
                                    }
                                )
                            } catch (error) {
                                console.log(error)
                            }
                        }
                    }
                }
            )
        } catch (error) {
            console.log(error)
        }
        return this.loggedUser
    }

    public login = () => {
        try {
            this.auth0.authorize({})
        } catch (err) {
            console.log(err)
        }
    }

    public checkSession = async () => {
        try {
            await this.auth0.checkSession({}, (err, data) => {
                if (err) {
                    console.log(err)
                    this.logout()
                } else {
                    if (!data.accessToken) {
                        this.logout()
                    }
                }
            })
        } catch (err) {
            console.log(err)
        }
    }

    public resetTimer() {
        this.setTimer()
    }

    public logout = () => {
        if (this.checkSessionInterval) {
            clearInterval(this.checkSessionInterval)
        }
        localStorage.removeItem('user')
        localStorage.removeItem('idToken')
        localStorage.removeItem('expiresAt')
        this.loggedUser = undefined
        this.auth0.logout({ returnTo: this.settings.redirectUri })
    }

    private setTimer() {
        console.debug('logout timeout: ', this.logoutTimeout)
        clearInterval(this.logoutTimeout)
        this.logoutTimeout = setTimeout(() => {
            this.logout()
        }, LOGOUT_INTERVAL * 1000)
    }

    private setUser(user: Auth0UserProfile) {
        this.loggedUser = user
        localStorage.setItem('user', JSON.stringify(user))
    }

    private setVars(data: any) {
        if (data.expiresIn) {
            // Set the time that the Access Token will expire at
            const expiresAt = data.expiresIn * 1000 + new Date().getTime()
            // this.accessToken = data.accessToken
            try {
                axios.defaults.headers.common.Authorization = `Bearer ${
                    data.accessToken
                }` // for all requests
                this.setTimer()
            } catch (err) {
                console.log('impossible to parse Bearer Token')
            }
            this.idToken = data.idToken
            this.expiresAt = expiresAt
            localStorage.setItem('expiresAt', JSON.stringify(this.expiresAt))
            localStorage.setItem('idToken', JSON.stringify(this.idToken))
        }
    }
}
export default AuthStore
