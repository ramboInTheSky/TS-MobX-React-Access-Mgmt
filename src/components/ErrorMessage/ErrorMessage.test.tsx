import React from 'react'
import { shallow } from 'enzyme'

import { ErrorMessage, ErrorMessageProps } from './ErrorMessage'
import { RoutingMock } from '../../stores/mock'

describe('<ErrorMessage />', () => {
    let props: any
    const Component = (ErrorMessage as any).wrappedComponent
    beforeEach(() => {
        props = {
            routing: new RoutingMock(),
        }
    })
    it('renders correctly', () => {
        expect(shallow(<Component {...props} />)).toMatchSnapshot()
    })

    // it('passes on props correctly', () => {
    // 	expect(shallow(<ErrorMessage foo="bar">Hello</ErrorMessage>)).toMatchSnapshot()
    // })

    // it('Does not re-render on children props change', () => {
    // 	const el: any = shallow(<ErrorMessage foo="bar" />)
    // 	el.setProps({ children: 'Goodbye' })
    // 	el.update()
    // 	expect(el).toMatchSnapshot()
    // })
})
