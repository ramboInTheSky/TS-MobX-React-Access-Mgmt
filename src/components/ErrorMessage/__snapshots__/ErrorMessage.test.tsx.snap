// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorMessage /> renders correctly 1`] = `
<Styled(div)>
  <h3>
    Something has gone wrong!
  </h3>
  <br />
  <br />
  <Link
    onClick={[Function]}
    to="/"
  >
    Go back to home page
  </Link>
</Styled(div)>
`;
